File: 接话茬.py
ps = input();cs = input();xs = input()
nextc=[];nextx=[]
nextc.append(-1);i=1; j=0
while i < len(cs):
    if ps[i]==ps[j] or j==-1:
        i+=1;j+=1
        nextc.append(j)
    else:
        j = nextc[j]
nextx.append(-1);i=1; j=0
while i < len(xs):
    if ps[i]==ps[j] or j==-1:
        i+=1;j+=1
        nextx.append(j)
    else:
        j = nextx[j]
i=0; j=0
if len(ps) > len(cs): i = len(ps)-len(cs)
while i < len(ps):
    if ps[i]==cs[j]:
        i+=1;j+=1;
    elif ps[i]!=cs[j] and j==0:
        i+=1
    else:#如果不相等时，判断
        j = nextc[j]
print(j,end=' ')
i=0; j=0
if len(ps) > len(xs): i = len(ps)-len(xs)
while i < len(ps):
    if ps[i]==xs[j]:
        i+=1;j+=1;
    elif ps[i]!=xs[j] and j==0:
        i+=1
    else:
        j = nextx[j]
print(j)

File: 接话茬.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <iostream>
using namespace std;
#define MAXSIZE 1000000
void GetNext(char cps[], int cpn[])
{
    int i=1, j=0, cpsn;
    cpn[0] = -1;
    cpsn = strlen(cps);
    while(i < cpsn)
    {
        if(cps[i]==cps[j] || j==-1)
        {
            i++;j++;
            cpn[i] = j;
        }
        else
        {
            j = cpn[j];
        }
    }

}
int GetResult(char* ps1, char* ps2, int* pn)
{
    int i=0, j=0;
    int ps1n, ps2n;
    ps1n = strlen(ps1);ps2n = strlen(ps2);
    if(ps1n > ps2n)
        i = ps1n-ps2n;
    while(i < ps1n)
    {
        if(ps1[i] == ps2[j])
            {i++;j++;}
        else if(ps1[i]!=ps2[j] && j==0)
            i++;
        else
            j = pn[j];
    }
    return j;
}
char ps[MAXSIZE],cs[MAXSIZE],xs[MAXSIZE];
int nextc[MAXSIZE],nextx[MAXSIZE];
int main()
{

    cin >> ps >> cs >> xs;
    GetNext(cs,nextc);
    GetNext(xs,nextx);
    int a, b;
    a = GetResult(ps, cs, nextc);
    b = GetResult(ps, xs, nextx);
    cout << a << " " << b << endl;
    return 0;
}



